src/main/java/com/conveyal/taui/controllers/AggregationAreaController.java

```diff
@@ -99,47 +119,79 @@ public static AggregationArea createAggregationArea (Request req, Response res)
-                    .map(f -> (Geometry) f.getDefaultGeometry())
-                    .collect(Collectors.toList());
-        UnaryUnionOp union = new UnaryUnionOp(geometries);
-        Geometry merged = union.union();
 
-        Envelope env = merged.getEnvelopeInternal();
-        Grid maskGrid = new Grid(SeamlessCensusGridExtractor.ZOOM, env.getMaxY(), env.getMaxX(), env.getMinY(), env.getMinX());
+        // 2. Read features ============================================================================================
+        List<SimpleFeature> features = reader.wgs84Stream().collect(Collectors.toList());
+        Map<String, Geometry> areas = new HashMap<>();
 
-        // Store the percentage each cell overlaps the mask, scaled as 0 to 100,000
-        List<Grid.PixelWeight> weights = maskGrid.getPixelWeights(merged, true);
-        weights.forEach(pixel -> maskGrid.grid[pixel.x][pixel.y] = pixel.weight * 100_000);
-
-        ObjectMetadata metadata = new ObjectMetadata();
-        metadata.setContentEncoding("gzip");
-        metadata.setContentType("application/octet-stream");
-
-        AggregationArea aggregationArea = new AggregationArea();
-        aggregationArea.name = maskName;
-        aggregationArea.regionId = regionId;
-
-        // Set `createdBy` and `accessGroup`
-        aggregationArea.accessGroup = req.attribute("accessGroup");
-        aggregationArea.createdBy = req.attribute("email");
-
-        File gridFile = new File(tempDir, "weights.grid");
-        OutputStream os = new GZIPOutputStream(new BufferedOutputStream(new FileOutputStream(gridFile)));
-        maskGrid.write(os);
-        os.close();
-
-        // Create the aggregation area before generating the S3 key so that the `_id` is generated
-        Persistence.aggregationAreas.create(aggregationArea);
-
-        InputStream is = new BufferedInputStream(new FileInputStream(gridFile));
-        // can't use putObject with File when we have metadata . . .
-        S3Util.s3.putObject(AnalysisServerConfig.gridBucket, aggregationArea.getS3Key(), is, metadata);
-        is.close();
+        if (features.size() > MAX_FEATURES) {
+            throw AnalysisServerException.fileUpload(MessageFormat.format("The uploaded shapefile has {0} features, " +
+                    "which exceeds the limit of {1}", features.size(), MAX_FEATURES));
+        }
 
-        tempDir.delete();
+        if (Boolean.parseBoolean(req.params("union"))) {
+            // Union (single combined aggregation area) requested
+            List<Geometry> geometries = features.stream().map(f -> (Geometry) f.getDefaultGeometry()).collect(Collectors.toList());
+            UnaryUnionOp union = new UnaryUnionOp(geometries);
+            // Name the area using the name in the request directly
+            areas.put(maskName, union.union());
+        } else {
+            // Don't union. Name each area by looking up its value for the name property in the request.
```

Comment:
Personally I'd rather not have parameters change meaning depending on other parameters (union=true interprets `name` as the name of the unioned feature, rather than the name of a property on features in the uploaded file). I would probably make them two different query parameters (e.g. `maskName` and `maskNameProperty`. But this is a detail.